### PR TITLE
sync_struct: Add {add,remove}_notifier()

### PR DESCRIPTION
**Pull request by dnadlinger**
Being able to dynamically (un)publish objects is essential for
some use cases; previously this could only be done by creating
a new Publisher (and thus TCP port) for each dynamically added
notifier, with obvious shortcomings.

This has been battle-tested in production for a considerable
amount of time in the Oxford group as part of an ndscan-related
ARTIQ master extension.

@sbourdeauducq: Good to merge?

### CODE REVIEW

File: sipyco/sync_struct.py [hunk 1]
```diff
@@ -319,11 +361,19 @@ def __init__(self, notifiers):
...
                     writer.write(line)
                     # raise exception on connection error
                     await writer.drain()
             finally:
-                self._recipients[notifier_name].remove(queue)
+                try:
+                    self._recipients[notifier_name].remove(queue)
+                except KeyError:
+                    # Notifier might have been removed using remove_notifier
+                    # already.
```
Comment (sbourdeauducq) [inline sipyco/sync_struct.py]: Won't this break if you remove a notifier, then add it again immediately with the same name?

Comment (dnadlinger) [reply sipyco/sync_struct.py]: Hm, what would we expect for existing connections in that scenario? (This just deals with the one connection queue, doesn't it?) I think closing them would be fine, although, granted, we are never re-adding listeners in our use cases.

Comment (sbourdeauducq) [reply sipyco/sync_struct.py]: What is the use scenario exactly?

**Comment by hartytp** [comment]
@dnadlinger @sbourdeauducq what's blocking this from being merged right now? This is one of the annoying points of divergence that makes it hard to use ndscan right now...

**Comment by sbourdeauducq** [comment]
> @dnadlinger @sbourdeauducq what's blocking this from being merged right now?

Complex patch, no time.

**Comment by Hanros94** [comment]
how do I apply this patch?

File: sipyco/sync_struct.py (hunk 1 above)
Comment (hartytp) [reply sipyco/sync_struct.py:375]: @dnadlinger will be better-placed to comment than me, but my (limited) understanding is that this is an essential feature for ndscan.

One of the really nice things about ndscan is that the applets know a lot about the scan. So, e.g. one has a single applet which can handle scanning an arbitrary number of dimensions, can set up axes correctly, etc without manual intervention (which is not possible in artiq). There are also really nice features like being able to update a dataset from scan results by right clicking in an applet.

The scans needs to be able to run in parallel, interrupting each other.

The way the scans are synchronised with the applets is using a separate notifier for each RID (namespaced datasets). Then using datasets to broadcast metadata. https://github.com/OxfordIonTrapGroup/ndscan/blob/7dd4a9de5255bf5a92a38750e020d0eaa7b17385/ndscan/experiment/entry_point.py#L445-L472

**Comment by charlesbaynham** [comment]
@sbourdeauducq it'd be good to see this merged if possible. `ndscan` is pretty well tested, so I think it's unlikely there are serious bugs remaining here.

**Comment by sbourdeauducq** [comment]
What about this: https://github.com/m-labs/sipyco/pull/3#discussion_r359137998

**Comment by hartytp** [comment]
It sounds like the overall [plan](https://chat.m-labs.hk/m-labs/pl/hog6bo4pniy8pemzzfkagpg8oy) is to move ndscan away from dataset namespaces, in which case this patch wouldn't be needed.

**Comment by hartytp** [comment]
But we would still need to agree on a solution which doesn't result in unworkable numbers of warnings generated through multiple dataset access.

**Comment by dnadlinger** [comment]
We now no longer use the local ARTIQ patch that relied on this.

**Pull request closed without merging**